### PR TITLE
feat(dapps) implement wallet connect URI input

### PR DESCRIPTION
**Pull request by stefandunca**
#### Other changes:

`HEAD~1`

- start Wallet Connect SDK using web engine after wallet is ready
- source WC project ID from the constants in nim
- close list when opening the `ConnectDappModal`
- remove the old POC modal instantiation
- rename ConnectDappPopup to Modal for consistency
- move specific dapps related user settings to their place in the component
- add DAppsStore to be used for dapps abstraction

`HEAD` - storybook

- add `ConnectDappModal` to story book
- mock FeatureFlags nim to silence `Global.qml` warnings

| Scope | Implementation | Design |
| - | - | - |
| Ready to paste | image | image |
| Pasted URI | image | image |
| Invalid URI | image | image |

### CODE REVIEW

File: storybook/pages/ConnectedDappsButtonPage.qml [hunk 1]
```diff
@@ -56,10 +59,33 @@ Item {
         ColumnLayout {
             id: optionsSpace
 
+            RowLayout {
+                Text { text: "projectId" }
+                Text {
+                    id: projectIdText
+                    readonly property string projectId: SystemUtils.getEnvVar("WALLET_CONNECT_PROJECT_ID")
```
Comment (caybro) [inline storybook/pages/ConnectedDappsButtonPage.qml:66]: Maybe it's a good idea to move `SystemUtils` out of SB into StatusQ, wdyt?

**Review by caybro — COMMENTED** [review]
LGTM in general

File: ui/imports/shared/popups/walletconnect/ConnectDappModal.qml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+import "ConnectDappModal"
+
+StatusModal {
+    id: root
+
+    signal pair(string uri)
+
+    width: 480
+    height: 633
```
[the remark this answers is not in the document]
Comment (stefandunca) [reply ui/imports/shared/popups/walletconnect/ConnectDappModal.qml:19]: I changed to `implicitHeight` as per design.

File: ui/imports/shared/popups/walletconnect/ConnectDappModal.qml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        StatusBaseText {
+            text: "WalletConnect URI"
+        }
+
+        WCUriInput {
+            id: uriInput
+        }
+
+        // Spacer
+        ColumnLayout {}
```
[the remark this answers is not in the document]
Comment (stefandunca) [reply ui/imports/shared/popups/walletconnect/ConnectDappModal.qml:38]: Using alignment I can't get it to stick property to top. Therefore, I use a Layout that expands verticall as a "spacer" to make it take the middle space and push the top component to top based on margins and padding and the bottom components to bottom. Any recommendations?

| Layout.alignment | Spacer (Layout) |
| - | - |
| image | image |

File: storybook/pages/ConnectedDappsButtonPage.qml (hunk 1 above)
Comment (stefandunca) [reply storybook/pages/ConnectedDappsButtonPage.qml:66]: That will be good to be done when needed. However given that I don't see a need to have it in the app (via StatusQ or use it in the library) I postpone this for later.

File: ui/imports/shared/popups/walletconnect/ConnectDappModal.qml (hunk 3 above)
[the remark this answers is not in the document]
Comment (caybro) [reply ui/imports/shared/popups/walletconnect/ConnectDappModal.qml:38]: I think what you want it:
```qml
Item { Layout.fillHeight: true }
```

**Comment by micieslak** [comment]
I'm looking at the pr and I have some suggestions to the storybook
please wait for my full review before merging

**Review by micieslak — CHANGES_REQUESTED** [review]
Hey! I took a closer look and came up with some conclusions.

1. In the real app we use .qrc file and compile qml and related sources into binary files. However in Storybook we load files directly from file system, what is important because of hot reloading mechanism. For the same sources both methods works transparently and fine for both cases.

The problem starts when we try to explicitly use one mechanism, like in `WalletConnectSDK`:
```
WebEngineLoader {
  // ...
  url: "qrc:/imports/shared/popups/walletconnect/sdk/src/index.html"
}
```
In this case it won't work in Storybook bc this mechanism is not used there.

However it can be easily solved by using relative path:
```
url: "sdk/src/index.html"
```
Qt will used the same mechanism as used for file where the path is used. So it will be via qrc in the app and file system in the Storybook.

Nothing more is needed then, paths should be resolved properly as for all other files. No need to use workarounds like `stubs/stub-resources.qrc`.

2. Stores are intended to be cut-off line for nim backend. Stubs are defined to override original import paths for stores with empty objects. Those empty objects provide matching type name and full flexibility for mocking, independently in every storybook's page. For that reason stubs should be just
```
import QtQml 2.15

QtObject {}
```
in a Storybook page store's content can be easily mocked e.g. this way:
```
    DAppsStore {
        readonly property WalletConnectSDK wCSDK: WalletConnectSDK {
            active: true

            projectId: projectIdText.projectId

            onSessionRequestEvent: (details) => {
                // TODO #14556
                console.debug(`@dd onSessionRequestEvent: ${JSON.stringify(details)}`)
            }
        }
    }
```
3. As mentioned above, stores are cut-off line between UI and backend. It should be thin layer exposing backend stuff. With this assumption agreed, we shouldn't keep components like `WalletConnectSDK` in a store, because it's not backend's component. When `WalletConnectSDK` is kept in a store, in Storybook we need to "mock" `wCSDK` with the real implementation, same as in a real impl.

4. I'm not sure what's the goal of `FeatueFlags.qml` in `stubs` dir. There is no usages of that. However it shoudn't be there anyways, because `stubs` is intended to keep store placeholders.

5. This point is not directly part of this PR, but still somehow worth fixing at early stage:
Right now we have:
```
readonly property var featureFlags: featureFlagsRootContextProperty
```
This code will give errors like `ReferenceError: featureFlagsRootContextProperty is not defined` in Storybook and Qml tests because there is backend's context property access not hidden behind stores (which are replaced by stubs in sb/tests env).

In some places in code we have similar pattern and to suppress the error we use following workaround:
```
property var mainModuleInst: typeof mainModule !== "undefined" ? mainModule : null
```
however, it's still far from ideal situation where that kind of access is encapsulated in stores.
For that reason that access to `featureFlagsRootContextProperty` should be placed in a store. It can be existing store or also new one, like `FeatureFlagsStore`. It could be top-level, independent store.
The drawback is that in `WalletHeader` where we need that flag, we cannot take it directly from global state - we need to declare it as explicit dependency (like bool property dappsEnabled) in that component and components above. It's a bit more to cumbersome but also more explicit, making components dependencies easier to track and understand. 

Please take those points into consideration. I'm open for discussion about that, it's important for app long-term stability to be on the same page about core architectural patterns that we follow. Thanks!

**Comment by stefandunca** [comment]
Thanks @micieslak for the great input. The proposed improvements are not the scope of this intermediate PR so I propose to merge this as it is and follow up with this improvements later on as described in the inline reply. What do you think?

> 1. In the real app we use .qrc file and compile qml and related sources into binary files. However in Storybook we load files directly from file system, what is important because of hot reloading mechanism. For the same sources both methods works transparently and fine for both cases.
> ...
>   url: "qrc:/imports/shared/popups/walletconnect/sdk/src/index.html"

I was thinking the same with the first POC implementation that was indeed loading the SDK from disk in storybook. That served me very well in quickly testing and prototyping with the code. However, @saledjenic removed that later on when the SDK was more stable and there was need for simplicity against flexibility, therefore to quickly have the current storybook work with the SDK and help in quick development I added the `stub-resources.qrc` that loads the SDK in the same way for storybook as for the application to be useful for manual testing. Let's reconsider this if really needed later on after we have more experience and input as the development advances.

> 2. Stores are intended to be cut-off line for nim backend. Stubs are defined to override original import paths for stores with empty objects. Those empty objects provide matching type name and full flexibility for mocking, independently in every storybook's page. For that reason stubs should be just
> 
> ```
> import QtQml 2.15
> 
> QtObject {}
> ```

I agree with you point here that is why in my followup implementation I already did this, see [DAppsStore.qml in #14615](https://github.com/status-im/status-desktop/blob/8e79216864d7851d295b7671b523b0fb73b1b482/storybook/stubs/shared/stores/DAppsStore.qml) by bringing the dapp service module and reverse the dependency.

> 3. As mentioned above, stores are cut-off line between UI and backend. It should be thin layer exposing backend stuff. With this assumption agreed, we shouldn't keep components like `WalletConnectSDK` in a store, because it's not backend's component. When `WalletConnectSDK` is kept in a store, in Storybook we need to "mock" `wCSDK` with the real implementation, same as in a real impl.

As mentioned above in my current work ([#14615](https://github.com/status-im/status-desktop/pull/14615)) I already improved on this point and will continue to improve with each change. The current PR was focusing on bringing the connect modal and have minimum supporting changes.

> 4. I'm not sure what's the goal of `FeatueFlags.qml` in `stubs` dir. There is no usages of that. However it shoudn't be there anyways, because `stubs` is intended to keep store placeholders.

That is to silence warning by providing a stub for the context property. I totally agree with your point but I could not see a quick way to have that component available globally as required by feature flags. At the moment I don't have a good understanding of the stores dependency and understand which is a common store for all the feature stores so I did this quick nim mock in order to progress on the current implementation.

> 5. This point is not directly part of this PR, but still somehow worth fixing at early stage:
>    Right now we have:
> 
> ```
> readonly property var featureFlags: featureFlagsRootContextProperty
> ```
> 
> This code will give errors like `ReferenceError: featureFlagsRootContextProperty is not defined` in Storybook and Qml tests because there is backend's context property access not hidden behind stores (which are replaced by stubs in sb/tests env).

The nim stub change in this PR fixes this and, as discussed in private, having a common store for exposing feature flags is the way to go in the future, however this is a separate topic that I would like not to mix with the current implementation. As discussed with @noeliaSD your team can help here given that it is a general implementation for all the upcoming features and not to wait this to become a priority for the dapp team.

File: ui/imports/shared/popups/walletconnect/ConnectDappModal.qml (hunk 3 above)
[the remark this answers is not in the document]
Comment (stefandunca) [reply ui/imports/shared/popups/walletconnect/ConnectDappModal.qml:38]: I was relying on the fact that `Layout`s have fill properties `true` by default. I will change to using `Item` with explicit `fillHeight` if considered to be more intuitive.

**Comment by stefandunca** [comment]
@micieslak following our discussion I removed the `stub-resources.qrc` and will come with a separate PR to follow on your proposal captured here: https://github.com/status-im/status-desktop/issues/14696

**Comment by micieslak** [comment]
> @micieslak following our discussion I removed the `stub-resources.qrc` and will come with a separate PR to follow on your proposal captured here: #14696

Hey Stefan, thank you for creating the follow-up tasks. I played a bit with relative paths and qrc files. And you are right that the situation is complicated bc some resources are from the app and some are from StatusQ, therefore is probably impossible to handle the uniformly from a single html file. So if it's necessary I'm fine with adding additional `.qrc` file as a workaround in Storybook. The only suggestion here is to keep it separately outside `stub` directory (which purpose is different), maybe just on top-level, and add proper explanation in `CMakeLists.txt`, where it's listed.

Regarding ctx props mocks, that's nice idea to have this mechanism until we full enclose access to backend in stores. Only for clarity I would move them to a separate dir, outside stubs: https://github.com/status-im/status-desktop/issues/14726

I also created a task for something that I should have done a long time ago - Readme for Storybook: https://github.com/status-im/status-desktop/issues/14724

**Review by micieslak — APPROVED** [review]

**Pull request merged**